text/0000-app-image-extensions.md

```diff
@@ -21,48 +22,139 @@ Allowing buildpacks to install OS packages dynamically during build would drasti
 However, it is not uncommon for application code to depend on OS packages.
 
-Given that app authors and platform maintainers experience increased build time directly, extending the stack at the app author's or platform maintainer's request may add flexibility without sacrificing performance.
+Given that app authors and platform maintainers experience increased build time directly, extending the stack at the app author's request may add flexibility without sacrificing performance. At the same time, we want the CNB interface to be highly orthogonal, which is why we've strived to "make everything a buildpack". Allowing some buildpacks to run with privileges would give users the flexibility they expect based on their experience with other tools like `Dockerfile`.
+
+The advantages of using a buildpack for privileged operations are the same as for unprivileged operations: they are composable, fast (caching), modular, reuseable, and safe.
+
+As an example, consider the use case of Acme.com. They have three teams that want to contibute privileged layers to every image built at the company. Each of these teams wants to manage their own artifacts and the mechanisms for installing them, but none of these teams are in control of the base image. The teams have special logic for when their layers should be added to an image; in some cases it's the precense of a Node.js app, and in others it's the use of Tomcat, etc. In the past, these teams were trying to contribute their layers by selectively adding them to every `Dockerfile` for every base image in the company, but this doesn't scale well. With Root Buildpacks, each team can manage their own artifacts, release cadence, and add logic to `bin/detect` to ensure the layers are added when needed.
 
 # What it is
 [what-it-is]: #what-it-is
 
-## Specification
+- *root buildpack* - a special case of buildpack that is run with privileges
+
+Application developers can use root buildpacks to extend their build and/or run images.
+
+We introduce a boolean `privileged` key in the `[buildpack]` table of `buildpack.toml`, which is defined as follows:
+
+```
+[buildpack]
+privileged = <boolean (default=false)>
+```
+
+When `privileged` is set to `true`, the lifecycle will run this buildpack as the `root` user.
 
-Stack image creators may add the following executables to their run and/or build images:
+For each root buildpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the buildpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). All of the captured changes will be included in a single layer produced as output from the buildpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
 
-Extension:
+The buildpack can then exclude directories from the layer by writing to the `launch.toml` using a new `[[excludes]]` table, for example:
+
+```toml
+[[excludes]]
+paths = [ "/var" ]
```

Comment:
Does this mean that root buildpack authors need to know what paths the packages the buildpack installs write to for layer exclusion?

Reply:
If they want to slim down the run image, then yes they do. Otherwise they can just ignore this file. This is a good thing to point out because this proposal is tilting toward giving the buildpack authors more responsibility. But with that, I think we keep a highly orthogonal interface (i.e. everything is a buildpack).

Reply:
Yeah just thinking allowed. I think that might work well for cases where a root buildpack is specific to a particular package - say, installing curl - but not so well if they support arbitrary packages based on files in the application repo?